src/librustc_codegen_llvm/common.rs

```diff
@@ -393,3 +400,186 @@ pub fn struct_in_context(
 fn hi_lo_to_u128(lo: u64, hi: u64) -> u128 {
     ((hi as u128) << 64) | (lo as u128)
 }
+
+pub fn add_define_metadata<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>,
+                                      instance: Instance<'tcx>,
+                                      llfn: &'ll Value) {
+    if cx.sess().opts.debugging_opts.call_metadata  {
+        if let Some(trait_refs) = cx.tcx.drop_glue(instance) {
+            add_define_drop_metadata(cx, &trait_refs, llfn);
+        } else {
+            let dyn_ = instance.trait_ref_and_method(cx.tcx).and_then(|(trait_ref, method)| {
+                if cx.tcx.dynamic_dispatch(trait_ref) {
+                    Some((trait_ref, method))
+                } else {
+                    None
+                }
+            });
+
+            let fn_ = if cx.tcx.function_pointer(instance) {
+                let sig = cx.tcx.normalize_erasing_late_bound_regions(
+                    ty::ParamEnv::reveal_all(),
+                    &instance.fn_sig(cx.tcx),
+                );
+                Some(sig)
+            } else {
+                None
+            };
+
+            if dyn_.is_some() || fn_.is_some() {
+                add_define_dyn_fn_metadata(cx, dyn_, fn_, llfn);
+            }
+        }
+    }
+}
+
+fn add_define_dyn_fn_metadata<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>,
+                                         dyn_: Option<(ExistentialTraitRef<'tcx>, Symbol)>,
+                                         fn_: Option<FnSig<'tcx>>,
+                                         llfn: &'ll Value) {
+    unsafe {
```

Comment:
the `llvm::*` functions here should have safe wrappers on `CodegenCx` instead of requiring this code to use unsafe so freely.